### PR TITLE
New syntax parsing

### PR DESCRIPTION
**Pull request by lfportal**
Introduces a new syntax for Spot. This PR only includes parsing of the raw data. Complex validation logic is out of scope.

### CODE REVIEW

File: libnext/src/models/nodes.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+export interface RequestNode {
+  headers: HeaderNode[];
+  pathParams: PathParamNode[];
+  queryParams: QueryParamNode[];
+  body?: BodyNode;
+}
+
+export interface ResponseNode extends DefaultResponseNode {
+  status: number;
```
Comment (fwouts) [inline libnext/src/models/nodes.ts:40]: Why not `status?` in `DefaultResponseNode` and just have one type?

File: libnext/src/models/types/kinds.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+export enum Kind {
+  Null = "null",
```
Comment (fwouts) [inline libnext/src/models/types/kinds.ts]: Styling note: personally, I prefer enum cases to be `UPPERCASE` as it makes it easier to separate them from types. Up to you though.

File: libnext/src/models/types/primitive-types.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import { Kind } from "./kinds";
+
+/** Primitive types native to the typescript language */
```
Comment (fwouts) [inline libnext/src/models/types/primitive-types.ts]:
```suggestion
/** Primitive types native to the TypeScript language. */
```

File: libnext/src/models/types/special-types.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    kind: Kind.Union,
+    types: unionTypes
+  };
+}
+
+export interface ReferenceType {
+  kind: Kind.TypeReference;
+  referenceKind: Kind;
+  name: string;
+  location: string;
```
Comment (fwouts) [inline libnext/src/models/types/special-types.ts:21]: Could you document this? What does a location look like?

File: libnext/src/models/types/kinds.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
+export enum Kind {
```
Comment (fwouts) [inline libnext/src/models/types/kinds.ts]: How about calling this `TypeKind`?

File: libnext/src/parsers/nodes/api-parser.spec.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import { parseApi } from "./api-parser";
+
+describe("@api parser", () => {
+  const apiClassName = "MyApi";
+  const apiName = "My API";
+
+  test("parses all information", () => {
+    const apiDescription = "Some description";
+    const content = `
+      /** ${apiDescription} */
```
Comment (fwouts) [inline libnext/src/parsers/nodes/api-parser.spec.ts]: I'd feel a bit more confident in these tests if we didn't use string interpolation (just hardcoding the constants, duplicating them here). I know that's less code reuse, but these are tests and we should try to minimise unnecessary logic as much as we can :)

File: libnext/src/parsers/nodes/default-response-parser.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ */
+export function parseDefaultResponse(
+  method: MethodDeclaration
+): DefaultResponseNode {
+  method.getDecoratorOrThrow("defaultResponse");
+  const headersParameter = methodParamWithDecorator(method, "headers");
+  const bodyParameter = methodParamWithDecorator(method, "body");
+  const description = extractJsDocComment(method);
+  const headers =
+    headersParameter === undefined ? [] : parseHeaders(headersParameter);
```
Comment (fwouts) [inline libnext/src/parsers/nodes/default-response-parser.ts]:
```suggestion
    headersParameter ? parseHeaders(headersParameter) : [];
```

File: libnext/src/models/nodes.ts (hunk 1 above)
Comment (lfportal) [reply libnext/src/models/nodes.ts:40]: Whilst it may be a bit trivial, it does make the internal representation a bit more typesafe

Comment (fwouts) [reply libnext/src/models/nodes.ts:40]: I think I understand why I was confused. `ResponseNode` is not semantically extending `DefaultResponseNode`: they're separate concepts. You're using `extends` as a convenient way to reuse code. I wonder if instead the code should be duplicated?

Either way, I think it's worth adding a bit of JSDoc to explain what each of these two types means.

**Review by fwouts — APPROVED** [review]
Nice work 🎉

**Pull request merged**